crypto/hpke/hpke_util.c

```diff
@@ -12,24 +12,200 @@
+     {OSSL_HPKE_KDFSTR_256, "0x1", "0x01", "1"}},
+    {OSSL_HPKE_KDF_ID_HKDF_SHA384,
+     {OSSL_HPKE_KDFSTR_384, "0x2", "0x02", "2"}},
+    {OSSL_HPKE_KDF_ID_HKDF_SHA512,
+     {OSSL_HPKE_KDFSTR_512, "0x3", "0x03", "3"}}
+};
+static synonymttab_t aeadstrtab[] = {
+    {OSSL_HPKE_AEAD_ID_AES_GCM_128,
+     {OSSL_HPKE_AEADSTR_AES128GCM, "0x1", "0x01", "1"}},
+    {OSSL_HPKE_AEAD_ID_AES_GCM_256,
+     {OSSL_HPKE_AEADSTR_AES256GCM, "0x2", "0x02", "2"}},
+    {OSSL_HPKE_AEAD_ID_CHACHA_POLY1305,
+     {OSSL_HPKE_AEADSTR_CP, "0x3", "0x03", "3"}}
+};
+
+/* Return an object containing KEM constants associated with a EC curve name */
+const OSSL_HPKE_KEM_INFO *ossl_HPKE_KEM_INFO_find_curve(const char *curve)
+{
+    int i;
+    int sz = OSSL_NELEM(hpke_kem_tab);
+
+    for (i = 0; i != sz; ++i) {
+        const char *group = hpke_kem_tab[i].groupname;
+
+        if (group == NULL)
+            group = hpke_kem_tab[i].keytype;
+        if (OPENSSL_strcasecmp(curve, group) == 0)
+            return &hpke_kem_tab[i];
+    }
+    ERR_raise(ERR_LIB_PROV, PROV_R_INVALID_CURVE);
+    return NULL;
+}
+
+const OSSL_HPKE_KEM_INFO *ossl_HPKE_KEM_INFO_find_id(uint16_t kemid)
+{
+    int i;
+    int sz = OSSL_NELEM(hpke_kem_tab);
+
+    for (i = 0; i != sz; ++i) {
+        if (hpke_kem_tab[i].kem_id == kemid)
+            return &hpke_kem_tab[i];
+    }
+    ERR_raise(ERR_LIB_PROV, PROV_R_INVALID_CURVE);
+    return NULL;
+}
+
+const OSSL_HPKE_KEM_INFO *ossl_HPKE_KEM_INFO_find_random(OSSL_LIB_CTX *ctx)
+{
+    unsigned char rval = 0;
+    int sz = OSSL_NELEM(hpke_kem_tab);
+
+    if (RAND_bytes_ex(ctx, &rval, sizeof(rval), 0) <= 0)
+        return NULL;
+    return &hpke_kem_tab[rval % sz];
+}
+
+const OSSL_HPKE_KDF_INFO *ossl_HPKE_KDF_INFO_find_id(uint16_t kdfid)
+{
+    int i;
+    int sz = OSSL_NELEM(hpke_kdf_tab);
+
+    for (i = 0; i != sz; ++i) {
+        if (hpke_kdf_tab[i].kdf_id == kdfid)
+            return &hpke_kdf_tab[i];
+    }
+    ERR_raise(ERR_LIB_PROV, PROV_R_INVALID_KDF);
+    return NULL;
+}
+
+const OSSL_HPKE_KDF_INFO *ossl_HPKE_KDF_INFO_find_random(OSSL_LIB_CTX *ctx)
+{
+    unsigned char rval = 0;
+    int sz = OSSL_NELEM(hpke_kdf_tab);
+
+    if (RAND_bytes_ex(ctx, &rval, sizeof(rval), 0) <= 0)
+        return NULL;
+    return &hpke_kdf_tab[rval % sz];
+}
+
+const OSSL_HPKE_AEAD_INFO *ossl_HPKE_AEAD_INFO_find_id(uint16_t aeadid)
+{
+    int i;
+    int sz = OSSL_NELEM(hpke_aead_tab);
+
+    for (i = 0; i != sz; ++i) {
+        if (hpke_aead_tab[i].aead_id == aeadid)
+            return &hpke_aead_tab[i];
+    }
+    ERR_raise(ERR_LIB_PROV, PROV_R_INVALID_AEAD);
+    return NULL;
+}
+
+const OSSL_HPKE_AEAD_INFO *ossl_HPKE_AEAD_INFO_find_random(OSSL_LIB_CTX *ctx)
+{
+    unsigned char rval = 0;
+    int sz = OSSL_NELEM(hpke_aead_tab);
+
+    if (RAND_bytes_ex(ctx, &rval, sizeof(rval), 0) <= 0)
+        return NULL;
+    /* the minus 1 below is so we don't pick the EXPORTONLY codepoint */
+    return &hpke_aead_tab[rval % (sz - 1)];
```

Comment:
You would normally do this in the int sz = () - 1;

Reply:
sure